File: finding-seams/kotlin/src/main/kotlin/findingseams/speakerfeedback/FeedbackSensor.kt
package findingseams.speakerfeedback

import kotlin.random.Random

class FeedbackSensor {
    /** Simulate an unreliable external device / service. */
    fun readScores(): List<Int>? {
        val outcome = Random.nextDouble()

        if (outcome < 0.05) {
            throw IllegalStateException("Sensor failure")
        }

        if (outcome < 0.1) {
            return null
        }

        val length = 1 + (Random.nextDouble() * 5).toInt()
        return List(length) { 1 + (Random.nextDouble() * 5).toInt() }
    }
}

File: finding-seams/csharp/FindingSeams/SpeakerFeedback/FeedbackSensor.cs
namespace FindingSeams.SpeakerFeedback;

public class FeedbackSensor
{
    public List<int>? ReadScores()
    {
        var outcome = Random.Shared.NextDouble();

        if (outcome < 0.05)
            throw new InvalidOperationException("Sensor failure");

        if (outcome < 0.1)
            return null;

        var length = 1 + (int)(Random.Shared.NextDouble() * 5);
        var scores = new List<int>(length);
        for (var i = 0; i < length; i++)
            scores.Add(1 + (int)(Random.Shared.NextDouble() * 5));

        return scores;
    }
}
